File: Data Structures & Algorithms/house-robber-ii/submission-3.py
class Solution:
    def solve(self, nums: List[int]) -> int:
        prev = 0
        curr = nums[0]
        for i in range(1, len(nums)):
            take = nums[i] + prev
            notake = curr
            prev = curr
            curr = max(take, notake)
        return curr
    def rob(self, nums: List[int]) -> int:
        if len(nums) == 1:
            return nums[0]
        return max(self.solve(nums[0:len(nums)-1]), self.solve(nums[1:]))

File: Data Structures & Algorithms/house-robber/submission-3.cpp
class Solution {
public:
    int rob(vector<int>& nums) {
        int prev = 0;
        int prev2 = nums[0];
        int ans = 0;
        for(int i = 1; i < nums.size(); i++){
            int take = nums[i];
            if(i > 1) take += prev;
            int noTake = prev2;
            ans = max(take, noTake);
            prev = prev2;
            prev2 = ans;
        } 
        return prev2;       
    }
};
